File: Movie/W10/Truck.java
package W10;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.StringTokenizer;

public class Truck {

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int W = Integer.parseInt(st.nextToken());
        int L = Integer.parseInt(st.nextToken());

        Deque<Integer> trucks = new ArrayDeque<>();

        st = new StringTokenizer(br.readLine());
        while (st.hasMoreTokens()) {
            trucks.addLast(Integer.parseInt(st.nextToken()));
        }

        Deque<Integer> bridge = new ArrayDeque<>();
        for (int i = 0; i < W; i++) {
            bridge.addLast(0);
        }

        int bridgeWeight = 0;

        Deque<Integer> finished = new ArrayDeque<>();
        int time = 0;

        while (finished.size() != N) {
            time++;

            if (bridge.peekFirst() != 0) {
                bridgeWeight -= bridge.peekFirst();
                finished.add(bridge.pollFirst());
            } else {
                bridge.pollFirst();
            }

            if (!trucks.isEmpty() && bridgeWeight + trucks.peekFirst() <= L) {
                bridgeWeight += trucks.peekFirst();
                bridge.addLast(trucks.pollFirst());
            } else {
                bridge.addLast(0);
            }

        }

        System.out.println(time);


    }
}

File: Jian/week10/truck.py
# 13335번 트럭

import sys
# queue 보다 사용 가능한 메서드가 더 많은 deque를 사용했다.
from collections import deque

input_line = sys.stdin.readline

n, w, L = list(map(int, input_line().rstrip().split()))
trucks = deque(list(map(int, input_line().rstrip().split())))
bridge = deque([0] * w)
time = 0

while not (sum(bridge) == 0 and not trucks):  # 다리가 비어 있지 않으면서 트럭 대기열에 트럭이 없을 때 까지
    bridge.popleft()
    time += 1
    if trucks and (trucks[0] + sum(bridge)) <= L:  # 다리에 새로운 truck이 추가 되어도 최대 하중 이하일 때
        bridge.append(trucks.popleft())
    else:
        bridge.append(0)


print(time)
